src/encryption/symmetric/aes/mod.rs

```diff
@@ -0,0 +1,295 @@
+pub type Block = [u8; 16];
+
+///  A word in AES represents a 32-bit array of data.
+pub type Word = [u8; 4];
+
+/// A generic N-bit key.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Key<const N: usize>
+where [(); N / 8]: {
+  inner: [u8; N / 8],
+}
+
+impl<const N: usize> Key<N>
+where [(); N / 8]:
+{
+  /// Creates a new `Key` of size `N` bits.
+  pub fn new(key_bytes: [u8; N / 8]) -> Self { Self { inner: key_bytes } }
+}
+
+impl<const N: usize> std::ops::Deref for Key<N>
+where [(); N / 8]:
+{
+  type Target = [u8; N / 8];
+
+  fn deref(&self) -> &Self::Target { &self.inner }
+}
+
+impl<const N: usize> SymmetricEncryption for AES<N>
+where [(); N / 8]:
+{
+  type Block = Block;
+  type Key = Key<N>;
+
+  /// Encrypt a message of size [`Block`] with a [`Key`] of size `N`-bits.
+  ///
+  /// ## Example
+  /// ```rust
+  /// #![feature(generic_const_exprs)]
+  ///
+  /// use rand::{thread_rng, Rng};
+  /// use ronkathon::encryption::symmetric::{
+  ///   aes::{Key, AES},
+  ///   SymmetricEncryption,
+  /// };
+  ///
+  /// let mut rng = thread_rng();
+  /// let key = Key::<128>::new(rng.gen());
+  /// let plaintext = rng.gen();
+  /// let encrypted = AES::encrypt(&key, &plaintext);
+  /// ```
+  fn encrypt(key: &Self::Key, plaintext: &Self::Block) -> Self::Block {
+    let num_rounds = match N {
+      128 => 10,
+      192 => 12,
+      256 => 14,
+      _ => panic!("AES only supports key sizes 128, 192 and 256 bits. You provided: {N}"),
+    };
+
+    Self::aes_encrypt(plaintext, key, num_rounds)
+  }
+
+  fn decrypt(_key: &Self::Key, _ciphertext: &Self::Block) -> Self::Block { unimplemented!() }
+}
+
+/// Contains the values given by [x^(i-1), {00}, {00}, {00}], with x^(i-1)
+/// being powers of x in the field GF(2^8).
+///
+/// NOTE: i starts at 1, not 0.
+const ROUND_CONSTANTS: [[u8; 4]; 10] = [
+  [0x01, 0x00, 0x00, 0x00],
+  [0x02, 0x00, 0x00, 0x00],
+  [0x04, 0x00, 0x00, 0x00],
+  [0x08, 0x00, 0x00, 0x00],
+  [0x10, 0x00, 0x00, 0x00],
+  [0x20, 0x00, 0x00, 0x00],
+  [0x40, 0x00, 0x00, 0x00],
+  [0x80, 0x00, 0x00, 0x00],
+  [0x1B, 0x00, 0x00, 0x00],
+  [0x36, 0x00, 0x00, 0x00],
+];
+
+/// A struct containing an instance of an AES encryption/decryption.
+#[derive(Clone)]
+pub struct AES<const N: usize> {}
+
+/// Instead of arranging its bytes in a line (array),
+/// AES operates on a grid, specifically a 4x4 column-major array:
+///
+/// [[b_0, b_4, b_8,  b_12],
+///  [b_1, b_5, b_9,  b_13],
+///  [b_2, b_6, b_10, b_14],
+///  [b_3, b_7, b_11, b_15]]
+///
+/// where b_i is the i-th byte. This is also how we will layout
+/// bytes in our `State`.
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
+struct State([[u8; 4]; 4]);
+
+impl<const N: usize> AES<N>
+where [(); N / 8]:
+{
+  /// Performs the cipher, with key size of `N` (in bits), as seen in Figure 5 of the document
+  /// linked in the front-page.
+  fn aes_encrypt(plaintext: &[u8; 16], key: &Key<N>, num_rounds: usize) -> Block {
```

Comment:
we can, but imo some of the asserts can/should coexist alongside `Result` usage, since they're asserting some of the invariants in the procedure (such as round keys being used up by the end of the encryption). Will see how I can refactor to have less `unwrap()`s.